### PR TITLE
overlord/state: do cancel in taskrunner goroutines that are waited but all waiting tasks are already in error

### PR DESCRIPTION
**Pull request by pedronis**
Also fix setting status when taskrunner and all goroutines are stopped, leaving it alone.

### CODE REVIEW

**Comment by pedronis** [comment]
needs #681

File: overlord/state/taskrunner.go [hunk 1]
```diff
@@ -78,16 +78,18 @@ func taskFail(t *Task) {
...
+	tb := &tomb.Tomb{}
+	r.tombs[task.ID()] = tb
+	tb.Go(func() error {
+		err := fn(task, tb)
 
 		r.state.Lock()
 		defer r.state.Unlock()
 		halted := task.HaltTasks()
-		if err == nil {
+		if err == nil && tb.Err() == tomb.ErrStillAlive {
```
Comment (niemeyer) [inline overlord/state/taskrunner.go]: That doesn't feel right. If something asks the tomb to stop for whatever reason, it won't return that, but it doesn't mean the task isn't done. What's the actual goal here?

File: overlord/state/taskrunner.go [hunk 2]
```diff
@@ -150,12 +166,15 @@ func (r *TaskRunner) Ensure() {
 			// a task can be in RunningStatus even when it
 			// is not started yet (like when the daemon
 			// process restarts)
-			if _, ok := r.tombs[t.ID()]; ok {
+			if tomb, ok := r.tombs[t.ID()]; ok {
+				if isPointless(t) {
```
Comment (niemeyer) [inline overlord/state/taskrunner.go]: Also don't understand this. The fact something it depends on is in error means nothing for this specific task. The stated dependency is the other way around.

**Comment by niemeyer** [comment]
This one branch feels a bit like a mystery to me. :-)

File: overlord/state/taskrunner.go (hunk 1 above)
Comment (niemeyer) [inline overlord/state/taskrunner.go]: Sorry, this comment is in the wrong location in the file changes pane. Please see the conversation pane for the right context (bug reported to GitHub).

File: overlord/state/taskrunner.go [hunk 3]
```diff
@@ -63,29 +63,45 @@ func (r *TaskRunner) Handlers() map[string]HandlerFunc {
...
+	tb := &tomb.Tomb{}
+	r.tombs[task.ID()] = tb
+	tb.Go(func() error {
+		err := fn(task, tb)
 
 		r.state.Lock()
 		defer r.state.Unlock()
-		if err == nil {
+		halted := task.HaltTasks()
+		if err == nil && tb.Err() == tomb.ErrStillAlive {
```
Comment (niemeyer) [inline overlord/state/taskrunner.go]: There are three cases which are being either mishandled by this logic or that make the implementation difficult to understand:

* The handler was asked to stop but succeeded doing its intended goal despite it
* The handler was asked to stop but would like to try again
* The handler would just like to try again later anyway, whether it was asked to stop or not.

We probably want something around these lines:
```
tb.Kill(fn(task, tb))
...
switch tb.Err() {
case Retry:
        // Do nothing. Handler asked to try again later.
case nil:
        task.SetStatus(DoneStatus)
        if len(task.HaltTasks()) > 0 {
                r.state.EnsureBefore(0)
        }
default:
        taskFail(task)
}
```

**Comment by pedronis** [comment]
addressing handling goroutine exits properly in a different PR.

the rest of this is an optimisation that can wait, will probably show up again when we have to undo things backward from a errored task through the chain of deps.

**Pull request closed without merging**